Tools/Scripts/webkitpy/layout_tests/servers/basic_dns_server.py

```diff
@@ -0,0 +1,51 @@
+# Copyright (C) 2023 Apple Inc. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from dnslib import CLASS, QTYPE, RR, RCODE
+from dnslib.server import BaseResolver, DNSServer
+
+
+class Resolver(BaseResolver):
+    def __init__(self, allowed_hosts=[]):
+        super().__init__()
+        self.hosts = list(map(lambda x: x if x.endswith(".") else x + ".", allowed_hosts))
+
+    def resolve(self, request, handler):
+        question = request.q
+        reply = request.reply()
+        if question.qtype == QTYPE.A and (question.qname in self.hosts or question.qname.matchSuffix("localhost")):
+            reply.add_answer(*RR.fromZone("{} 3600 A 127.0.0.1".format(question.qname)))
+        else:
+            reply.header.rcode = getattr(RCODE, "NXDOMAIN")
+        return reply
+
+
+if __name__ == "__main__":
+    import time
+    server = DNSServer(Resolver(["site.example"]), port=8053, address="127.0.0.1")
+    server.start_thread()
+    try:
```

Comment:
I'm a bit confused by this being threaded. Looks like the motivation here is to  be a stand-alone process (which I think is the right design!), I don't understand why we need the spin-loop here.

Reply:
This was/is only intended to be used for testing as a stand-alone process. When the service is run for WPT, it is only run as a background thread. The spin-loop just allows for using ^C and cleaning-up afterward when testing this script.

Reply:
I guess you might be asking why I chose this design instead of just calling `server.start()` (https://github.com/paulc/dnslib/blob/master/dnslib/server.py#L384). And I don't know/remember the answer to that - probably for mirroring the usage in the WPT code. I don't have a strong opinion here.

Reply:
If it's just for testing, mirroring the usage in WPT is correct. Probably worth noting that this is just intended for testing, though.